File: typescript/patterns/provider-discovery.ts
/**
 * Pattern: Provider Discovery and Selection
 *
 * Demonstrates how to discover, filter, and select providers
 * based on various criteria like price, reputation, and capabilities.
 *
 * This example shows:
 * - Discovering available providers for a service
 * - Filtering by price, reputation, capabilities
 * - Selection strategies (cheapest, best, fastest)
 * - Provider health checking
 * - Fallback provider chains
 *
 * Note: This example uses mock provider registry.
 * In production, this would query on-chain data.
 *
 * Run: npx ts-node patterns/provider-discovery.ts
 */

import { provide, request } from '@agirails/sdk';
import { waitForProvider } from '../src/utils/helpers';

// =========================================================
// TYPES
// =========================================================

interface ProviderInfo {
  address: string;
  service: string;
  price: number; // USDC per request
  reputation: number; // 0-100 score
  responseTimeMs: number; // Average response time
  capabilities: string[];
  availability: number; // 0-1 uptime ratio
  totalJobs: number;
  successRate: number; // 0-1
}

interface ProviderFilter {
  service: string;
  maxPrice?: number;
  minReputation?: number;
  maxResponseTime?: number;
  requiredCapabilities?: string[];
  minAvailability?: number;
  minSuccessRate?: number;
}

type SelectionStrategy = 'cheapest' | 'best' | 'fastest' | 'balanced';

// =========================================================
// MOCK PROVIDER REGISTRY
// =========================================================

/**
 * Simulated provider registry
 * In production, this would query blockchain/subgraph
 */
class ProviderRegistry {
  private providers: ProviderInfo[] = [
    {
      address: '0xProvider1...AAA',
      service: 'translation',
      price: 0.10,
      reputation: 95,
      responseTimeMs: 200,
      capabilities: ['en', 'de', 'fr', 'es'],
      availability: 0.99,
      totalJobs: 10000,
      successRate: 0.98,
    },
    {
      address: '0xProvider2...BBB',
      service: 'translation',
      price: 0.05,
      reputation: 75,
      responseTimeMs: 500,
      capabilities: ['en', 'de'],
      availability: 0.95,
      totalJobs: 500,
      successRate: 0.90,
    },
    {
      address: '0xProvider3...CCC',
      service: 'translation',
      price: 0.15,
      reputation: 90,
      responseTimeMs: 150,
      capabilities: ['en', 'de', 'fr', 'es', 'ja', 'zh'],
      availability: 0.999,
      totalJobs: 50000,
      successRate: 0.99,
    },
    {
      address: '0xProvider4...DDD',
      service: 'image-gen',
      price: 0.50,
      reputation: 88,
      responseTimeMs: 3000,
      capabilities: ['realistic', 'anime', 'abstract'],
      availability: 0.97,
      totalJobs: 8000,
      successRate: 0.95,
    },
    {
      address: '0xProvider5...EEE',
      service: 'image-gen',
      price: 0.25,
      reputation: 70,
      responseTimeMs: 5000,
      capabilities: ['realistic'],
      availability: 0.90,
      totalJobs: 200,
      successRate: 0.85,
    },
  ];

  /**
   * Get all providers for a service
   */
  async getProviders(service: string): Promise<ProviderInfo[]> {
    // Simulate network delay
    await new Promise((r) => setTimeout(r, 50));
    return this.providers.filter((p) => p.service === service);
  }

  /**
   * Get provider by address
   */
  async getProvider(address: string): Promise<ProviderInfo | null> {
    await new Promise((r) => setTimeout(r, 20));
    return this.providers.find((p) => p.address === address) || null;
  }
}

// =========================================================
// PROVIDER DISCOVERY
// =========================================================

class ProviderDiscovery {
  constructor(private registry: ProviderRegistry) {}

  /**
   * Discover providers matching filter criteria
   */
  async discover(filter: ProviderFilter): Promise<ProviderInfo[]> {
    console.log(`[Discovery] Searching for '${filter.service}' providers...`);

    // Get all providers for service
    let providers = await this.registry.getProviders(filter.service);
    console.log(`[Discovery] Found ${providers.length} providers`);

    // Apply filters
    if (filter.maxPrice !== undefined) {
      providers = providers.filter((p) => p.price <= filter.maxPrice!);
      console.log(`[Discovery] After price filter: ${providers.length}`);
    }

    if (filter.minReputation !== undefined) {
      providers = providers.filter((p) => p.reputation >= filter.minReputation!);
      console.log(`[Discovery] After reputation filter: ${providers.length}`);
    }

    if (filter.maxResponseTime !== undefined) {
      providers = providers.filter((p) => p.responseTimeMs <= filter.maxResponseTime!);
      console.log(`[Discovery] After response time filter: ${providers.length}`);
    }

    if (filter.requiredCapabilities?.length) {
      providers = providers.filter((p) =>
        filter.requiredCapabilities!.every((cap) => p.capabilities.includes(cap))
      );
      console.log(`[Discovery] After capabilities filter: ${providers.length}`);
    }

    if (filter.minAvailability !== undefined) {
      providers = providers.filter((p) => p.availability >= filter.minAvailability!);
      console.log(`[Discovery] After availability filter: ${providers.length}`);
    }

    if (filter.minSuccessRate !== undefined) {
      providers = providers.filter((p) => p.successRate >= filter.minSuccessRate!);
      console.log(`[Discovery] After success rate filter: ${providers.length}`);
    }

    return providers;
  }

  /**
   * Select best provider based on strategy
   */
  selectBest(providers: ProviderInfo[], strategy: SelectionStrategy): ProviderInfo | null {
    if (providers.length === 0) return null;

    console.log(`[Selection] Using '${strategy}' strategy...`);

    switch (strategy) {
      case 'cheapest':
        return providers.sort((a, b) => a.price - b.price)[0];

      case 'best':
        // Weighted score: reputation (40%) + success rate (40%) + availability (20%)
        return providers.sort((a, b) => {
          const scoreA = a.reputation * 0.4 + a.successRate * 100 * 0.4 + a.availability * 100 * 0.2;
          const scoreB = b.reputation * 0.4 + b.successRate * 100 * 0.4 + b.availability * 100 * 0.2;
          return scoreB - scoreA;
        })[0];

      case 'fastest':
        return providers.sort((a, b) => a.responseTimeMs - b.responseTimeMs)[0];

      case 'balanced':
        // Balance of price, quality, and speed
        return providers.sort((a, b) => {
          // Normalize each metric to 0-1 (lower is better for ranking)
          const priceScoreA = a.price / 1; // Assume max price $1
          const priceScoreB = b.price / 1;
          const qualityScoreA = 1 - a.reputation / 100;
          const qualityScoreB = 1 - b.reputation / 100;
          const speedScoreA = a.responseTimeMs / 5000; // Assume max 5s
          const speedScoreB = b.responseTimeMs / 5000;

          const totalA = priceScoreA * 0.33 + qualityScoreA * 0.33 + speedScoreA * 0.33;
          const totalB = priceScoreB * 0.33 + qualityScoreB * 0.33 + speedScoreB * 0.33;
          return totalA - totalB;
        })[0];

      default:
        return providers[0];
    }
  }

  /**
   * Check if provider is healthy
   */
  async checkHealth(provider: ProviderInfo): Promise<boolean> {
    console.log(`[Health] Checking ${provider.address.substring(0, 15)}...`);

    // Simulate health check (ping or test request)
    await new Promise((r) => setTimeout(r, 100));

    // Simulate 95% success rate for health checks
    const healthy = Math.random() > 0.05;
    console.log(`[Health] ${healthy ? 'Healthy' : 'Unhealthy'}`);
    return healthy;
  }
}

// =========================================================
// PROVIDER SELECTION WRAPPER
// =========================================================

interface SmartRequestOptions {
  service: string;
  input: any;
  budget: number;
  filter?: Partial<ProviderFilter>;
  strategy?: SelectionStrategy;
  fallbackStrategies?: SelectionStrategy[];
  healthCheck?: boolean;
}

/**
 * Make a request with smart provider selection
 */
async function smartRequest(
  discovery: ProviderDiscovery,
  options: SmartRequestOptions
) {
  const filter: ProviderFilter = {
    service: options.service,
    maxPrice: options.budget,
    ...options.filter,
  };

  // Discover providers
  const providers = await discovery.discover(filter);

  if (providers.length === 0) {
    throw new Error(`No providers found for '${options.service}' matching criteria`);
  }

  // Try strategies in order
  const strategies = [
    options.strategy || 'balanced',
    ...(options.fallbackStrategies || []),
  ];

  for (const strategy of strategies) {
    const selected = discovery.selectBest(providers, strategy);
    if (!selected) continue;

    console.log(`[Smart] Selected: ${selected.address.substring(0, 15)}... ($${selected.price})`);

    // Health check if enabled
    if (options.healthCheck) {
      const healthy = await discovery.checkHealth(selected);
      if (!healthy) {
        console.log(`[Smart] Provider unhealthy, trying next...`);
        // Remove from providers and try again
        const index = providers.indexOf(selected);
        if (index > -1) providers.splice(index, 1);
        continue;
      }
    }

    // Make request
    try {
      const { result, transaction } = await request(options.service, {
        input: options.input,
        budget: options.budget,
        // provider: selected.address, // Note: Provider selection would be used in production SDK
      });

      return {
        result,
        transaction,
        provider: selected,
      };
    } catch (error) {
      console.log(`[Smart] Request failed, trying next provider...`);
      const index = providers.indexOf(selected);
      if (index > -1) providers.splice(index, 1);
    }
  }

  throw new Error('All providers failed');
}

// =========================================================
// DEMO PROVIDERS
// =========================================================

async function startDemoProviders() {
  const translationProvider = provide('translation', async (job) => {
    const { text, targetLang } = job.input as { text: string; targetLang: string };
    await new Promise((r) => setTimeout(r, 200));
    return {
      original: text,
      translated: `[${targetLang.toUpperCase()}] ${text}`,
      targetLang,
    };
  });

  const imageGenProvider = provide('image-gen', async (job) => {
    const { prompt, style } = job.input as { prompt: string; style: string };
    await new Promise((r) => setTimeout(r, 500));
    return {
      prompt,
      style,
      imageUrl: `https://generated.example.com/${Date.now()}.png`,
      resolution: '1024x1024',
    };
  });

  await Promise.all([waitForProvider(translationProvider), waitForProvider(imageGenProvider)]);

  return {
    stop: async () => {
      await translationProvider.stop();
      await imageGenProvider.stop();
    },
  };
}

// =========================================================
// DEMO
// =========================================================

async function main() {
  console.log('='.repeat(60));
  console.log('AGIRAILS Pattern: Provider Discovery');
  console.log('='.repeat(60));
  console.log('');

  const registry = new ProviderRegistry();
  const discovery = new ProviderDiscovery(registry);
  const providers = await startDemoProviders();

  try {
    // Demo 1: Basic discovery
    console.log('--- Demo 1: Basic Provider Discovery ---\n');

    const translationProviders = await discovery.discover({
      service: 'translation',
    });

    console.log('\nAvailable translation providers:');
    translationProviders.forEach((p) => {
      console.log(`  ${p.address.substring(0, 15)}... | $${p.price} | Rep: ${p.reputation} | ${p.responseTimeMs}ms`);
    });
    console.log('');

    // Demo 2: Filtered discovery
    console.log('--- Demo 2: Filtered Discovery ---\n');

    const filteredProviders = await discovery.discover({
      service: 'translation',
      maxPrice: 0.12,
      minReputation: 80,
      requiredCapabilities: ['en', 'de', 'fr'],
    });

    console.log('\nFiltered providers (max $0.12, rep >= 80, supports en/de/fr):');
    filteredProviders.forEach((p) => {
      console.log(`  ${p.address.substring(0, 15)}... | $${p.price} | Rep: ${p.reputation}`);
      console.log(`    Caps: ${p.capabilities.join(', ')}`);
    });
    console.log('');

    // Demo 3: Selection strategies
    console.log('--- Demo 3: Selection Strategies ---\n');

    const allProviders = await discovery.discover({ service: 'translation' });

    const strategies: SelectionStrategy[] = ['cheapest', 'best', 'fastest', 'balanced'];
    for (const strategy of strategies) {
      const selected = discovery.selectBest(allProviders, strategy);
      if (selected) {
        console.log(`  ${strategy.toUpperCase()}: ${selected.address.substring(0, 15)}... ($${selected.price}, rep: ${selected.reputation}, ${selected.responseTimeMs}ms)`);
      }
    }
    console.log('');

    // Demo 4: Smart request with fallback
    console.log('--- Demo 4: Smart Request with Fallback ---\n');

    const result = await smartRequest(discovery, {
      service: 'translation',
      input: { text: 'Hello, world!', targetLang: 'de' },
      budget: 0.20,
      filter: { minReputation: 70 },
      strategy: 'best',
      fallbackStrategies: ['balanced', 'cheapest'],
      healthCheck: true,
    });

    console.log('\nResult:');
    console.log(`  Provider: ${result.provider.address.substring(0, 15)}...`);
    console.log(`  Price: $${result.provider.price}`);
    console.log(`  Translation: ${result.result?.translated || result.result}`);
    console.log('');

    // Summary
    console.log('='.repeat(60));
    console.log('Summary: Provider Discovery Patterns');
    console.log('='.repeat(60));
    console.log('');
    console.log('1. ProviderRegistry: Query available providers');
    console.log('   await registry.getProviders("service")');
    console.log('');
    console.log('2. ProviderDiscovery.discover(): Filter by criteria');
    console.log('   { maxPrice, minReputation, requiredCapabilities, ... }');
    console.log('');
    console.log('3. Selection strategies:');
    console.log('   - cheapest: Lowest price');
    console.log('   - best: Highest reputation + success rate');
    console.log('   - fastest: Lowest response time');
    console.log('   - balanced: Weighted combination');
    console.log('');
    console.log('4. smartRequest(): Auto-select with fallback');
    console.log('   Tries multiple strategies, skips unhealthy providers');
    console.log('');
    console.log('Best Practices:');
    console.log('  - Cache provider data (update periodically)');
    console.log('  - Health check before critical requests');
    console.log('  - Use fallback strategies for resilience');
    console.log('  - Consider provider specialization (capabilities)');
    console.log('');
  } finally {
    await providers.stop();
  }
}

main().catch(console.error);

File: python/patterns/provider_discovery.py
#!/usr/bin/env python3
"""
Pattern: Provider Discovery

Demonstrates how to discover, filter, and select the best provider
for a given service request.

This pattern shows:
- Service directory queries
- Provider filtering (price, reputation, availability)
- Best provider selection strategies
- Fallback mechanisms

Run: python patterns/provider_discovery.py
"""

import asyncio
import sys
import random
from pathlib import Path
from dataclasses import dataclass
from typing import List, Optional, Callable
from enum import Enum

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state, log, log_section


class SelectionStrategy(Enum):
    CHEAPEST = "cheapest"
    BEST_REPUTATION = "best_reputation"
    FASTEST = "fastest"
    ROUND_ROBIN = "round_robin"


@dataclass
class ProviderInfo:
    """Information about a provider."""
    address: str
    name: str
    services: List[str]
    price_per_request: float  # USDC
    reputation_score: float  # 0-100
    avg_response_time_ms: int
    success_rate: float  # 0-1
    is_available: bool


class ProviderDiscovery:
    """Discovers and selects providers for services."""

    def __init__(self):
        self.providers: List[ProviderInfo] = []
        self._round_robin_index = 0

    def register(self, provider: ProviderInfo) -> None:
        """Register a provider."""
        self.providers.append(provider)

    def discover(
        self,
        service: str,
        max_price: Optional[float] = None,
        min_reputation: Optional[float] = None,
        only_available: bool = True,
    ) -> List[ProviderInfo]:
        """Discover providers offering a service."""
        results = []

        for p in self.providers:
            # Check service
            if service not in p.services:
                continue

            # Check availability
            if only_available and not p.is_available:
                continue

            # Check price
            if max_price and p.price_per_request > max_price:
                continue

            # Check reputation
            if min_reputation and p.reputation_score < min_reputation:
                continue

            results.append(p)

        return results

    def select_best(
        self,
        providers: List[ProviderInfo],
        strategy: SelectionStrategy = SelectionStrategy.BEST_REPUTATION,
    ) -> Optional[ProviderInfo]:
        """Select the best provider using a strategy."""
        if not providers:
            return None

        if strategy == SelectionStrategy.CHEAPEST:
            return min(providers, key=lambda p: p.price_per_request)

        elif strategy == SelectionStrategy.BEST_REPUTATION:
            return max(providers, key=lambda p: p.reputation_score)

        elif strategy == SelectionStrategy.FASTEST:
            return min(providers, key=lambda p: p.avg_response_time_ms)

        elif strategy == SelectionStrategy.ROUND_ROBIN:
            selected = providers[self._round_robin_index % len(providers)]
            self._round_robin_index += 1
            return selected

        return providers[0]

    def select_with_fallback(
        self,
        service: str,
        strategies: List[SelectionStrategy],
        **filters,
    ) -> Optional[ProviderInfo]:
        """Try multiple strategies until one succeeds."""
        for strategy in strategies:
            providers = self.discover(service, **filters)
            if providers:
                selected = self.select_best(providers, strategy)
                if selected:
                    return selected

        # Last resort: any available provider
        all_providers = self.discover(service, only_available=True)
        return all_providers[0] if all_providers else None


async def main() -> None:
    log_section("AGIRAILS Pattern - Provider Discovery")

    clear_mock_state()

    # =====================================================
    # Part 1: Setup Mock Providers
    # =====================================================
    log_section("Part 1: Provider Registry")

    discovery = ProviderDiscovery()

    # Register mock providers
    providers_data = [
        ProviderInfo(
            address="0xAAAA" + "0" * 36,
            name="FastTranslate",
            services=["translate", "summarize"],
            price_per_request=0.50,
            reputation_score=85,
            avg_response_time_ms=200,
            success_rate=0.95,
            is_available=True,
        ),
        ProviderInfo(
            address="0xBBBB" + "0" * 36,
            name="QualityTranslate",
            services=["translate"],
            price_per_request=1.00,
            reputation_score=98,
            avg_response_time_ms=500,
            success_rate=0.99,
            is_available=True,
        ),
        ProviderInfo(
            address="0xCCCC" + "0" * 36,
            name="BudgetTranslate",
            services=["translate", "ocr"],
            price_per_request=0.25,
            reputation_score=70,
            avg_response_time_ms=800,
            success_rate=0.85,
            is_available=True,
        ),
        ProviderInfo(
            address="0xDDDD" + "0" * 36,
            name="OfflineProvider",
            services=["translate"],
            price_per_request=0.30,
            reputation_score=90,
            avg_response_time_ms=300,
            success_rate=0.92,
            is_available=False,
        ),
    ]

    for p in providers_data:
        discovery.register(p)
        status = "✅" if p.is_available else "❌"
        print(f"{status} {p.name}")
        print(f"   Services: {', '.join(p.services)}")
        print(f"   Price: ${p.price_per_request:.2f} | Rep: {p.reputation_score} | Speed: {p.avg_response_time_ms}ms")

    print()

    # =====================================================
    # Part 2: Basic Discovery
    # =====================================================
    log_section("Part 2: Basic Discovery")

    log("🔍", "Finding all 'translate' providers...")
    translate_providers = discovery.discover("translate")
    print(f"   Found: {len(translate_providers)} providers")
    for p in translate_providers:
        print(f"   - {p.name} (${p.price_per_request})")
    print()

    log("🔍", "Finding 'translate' providers under $0.50...")
    cheap_providers = discovery.discover("translate", max_price=0.50)
    print(f"   Found: {len(cheap_providers)} providers")
    for p in cheap_providers:
        print(f"   - {p.name} (${p.price_per_request})")
    print()

    log("🔍", "Finding 'translate' providers with reputation >= 90...")
    quality_providers = discovery.discover("translate", min_reputation=90)
    print(f"   Found: {len(quality_providers)} providers")
    for p in quality_providers:
        print(f"   - {p.name} (rep: {p.reputation_score})")

    # =====================================================
    # Part 3: Selection Strategies
    # =====================================================
    log_section("Part 3: Selection Strategies")

    all_translate = discovery.discover("translate")

    strategies = [
        (SelectionStrategy.CHEAPEST, "Cheapest"),
        (SelectionStrategy.BEST_REPUTATION, "Best Reputation"),
        (SelectionStrategy.FASTEST, "Fastest"),
    ]

    for strategy, name in strategies:
        selected = discovery.select_best(all_translate, strategy)
        if selected:
            print(f"{name}: {selected.name}")
            print(f"   Price: ${selected.price_per_request} | Rep: {selected.reputation_score} | Speed: {selected.avg_response_time_ms}ms")
        print()

    # Round robin
    log("🔄", "Round Robin (3 selections):")
    for i in range(3):
        selected = discovery.select_best(all_translate, SelectionStrategy.ROUND_ROBIN)
        print(f"   Selection {i + 1}: {selected.name}")

    # =====================================================
    # Part 4: Fallback Mechanism
    # =====================================================
    log_section("Part 4: Fallback Mechanism")

    log("🔍", "Looking for premium provider (rep >= 95, price <= $0.30)...")

    # This combination might not exist
    result = discovery.select_with_fallback(
        "translate",
        strategies=[
            SelectionStrategy.BEST_REPUTATION,
            SelectionStrategy.CHEAPEST,
        ],
        min_reputation=95,
        max_price=0.30,
    )

    if result:
        print(f"   Found: {result.name}")
    else:
        print("   No exact match, trying fallback...")

        # Relax constraints
        result = discovery.select_with_fallback(
            "translate",
            strategies=[SelectionStrategy.BEST_REPUTATION],
            min_reputation=80,
        )
        if result:
            print(f"   Fallback: {result.name} (relaxed constraints)")

    # =====================================================
    # Part 5: Integration with SDK
    # =====================================================
    log_section("Part 5: Integration with SDK")

    print("Usage with AGIRAILS SDK:")
    print()
    print("```python")
    print("# Discover best provider")
    print("providers = discovery.discover('translate', min_reputation=80)")
    print("best = discovery.select_best(providers, SelectionStrategy.BEST_REPUTATION)")
    print()
    print("# Use with request()")
    print("result = await request(")
    print("    'translate',")
    print("    input={'text': 'Hello'},")
    print("    budget=best.price_per_request * 1.1,  # 10% buffer")
    print("    provider=best.address,  # Specify provider")
    print(")")
    print("```")

    log("🎉", "Provider discovery demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
